### DISCUSSION TITLE
Keyboard shortcuts

### DISCUSSION THREAD

**Discussion by samejr** [apihero-run/jsonhero-web#59] [Ideas]
# Keyboard shortcuts – design and functionality 

I want to start a discussion about how we could implement nice keyboard shortcuts UI into JSON Hero. Below are some of the considerations so far and it'd be great to get some more input from people who use them regularly, on both PC and Mac platforms.

# How should the list of shortcuts look?

There seems to be a convention in a lot of products that show a clean, clear list of actions alongside the shortcut keys presented as key icons. Here are a few examples from various apps:

Invision's Freehand tool:
![CleanShot 2022-05-07 at 11 50 02](https://user-images.githubusercontent.com/7555566/167251807-71ac1073-9ae9-4cf4-b177-752421c6a1e1.png)

Linear:
<img width="348" alt="CleanShot 2022-05-07 at 08 34 31@2x" src="https://user-images.githubusercontent.com/7555566/167252030-a222875d-1724-4213-9777-c8df69c29796.png">

Figma:
![CleanShot 2022-05-07 at 12 19 54](https://user-images.githubusercontent.com/7555566/167252045-084e6e47-8999-46b1-9f64-1c578fb1eb7c.png)

Slack app:
<img width="399" alt="CleanShot 2022-05-07 at 12 33 19@2x" src="https://user-images.githubusercontent.com/7555566/167252449-fe14981f-4751-4760-8969-15e932cc3ebb.png">

Evernote app:
![CleanShot 2022-05-07 at 12 32 40](https://user-images.githubusercontent.com/7555566/167252457-20ab6778-d1c1-4a66-9681-8c133e6788c8.png)

# How would you access the list of shortcuts?

There a bunch of places a shortcuts panel could be accessed from. Here are the 2 of the more sensible spots IMO. Option 2. below is my favourite. Accessing the shortcuts panel would itself be accessed with a shortcut key of course.

![CleanShot 2022-05-07 at 11 47 57](https://user-images.githubusercontent.com/7555566/167251780-be72f198-2d26-457c-bce6-924b17866c70.png)

# Customising shortcuts

This seems to be typically a very nice but pro-feature not included in many popular web apps but I've included it here as some users might like to consider it as more of a priority. The functionality for editing is typically embedded into shortcuts panel itself. Here are some examples:

Cleanshot:
![CleanShot 2022-05-07 at 12 15 01](https://user-images.githubusercontent.com/7555566/167253504-a742657e-79d8-48e4-8214-e409c5bb04ec.png)

Xcode:
https://user-images.githubusercontent.com/7555566/167253430-2694fb4c-b2f0-4b77-ac4a-3a110ee2c647.mp4

VS Code:
https://user-images.githubusercontent.com/7555566/167253452-33652868-6071-4161-a678-83e28994b5f2.mp4

If anyone would like to take a look at a UI design or contribute to the UX discussion here, we are open to suggestions and contributions. 😄

**Comment by om-ray** [comment]
![Group 30](https://user-images.githubusercontent.com/38233712/167253141-72ab782e-dfca-4b28-9f67-d83577c0bce3.png)
This is the design I came up with for the keyboard shortcuts panel. It shows what shortcuts there are, the shortcut key, and an icon related to that key, with an input to change the keyboard shortcut on the side

**Comment by om-ray** [comment]
@samejr none of the above examples have a clear way that shows how we can change shortcuts and I think that's less than desirable from a UX perspective

**Reply by samejr** [reply]
Hey @123om123 thanks for posting your designs, they look great! A few more iterations should make for a great feature. I've made a few feedback notes below to hopefully help you with a version 2 design:

1. I think this feature should be triggered as a pop over from the main App page where the shortcuts are used rather than the Dashboard. My preferred spot is to replace the existing shortcut icons in the bottom left bar with a new 'Shortcuts' label along side the keyboard commands to open it. Likely something like Opt + S.
2. For the panel design itself, I think the 2-column examples above for Linear and Invision Freehand are the cleanest and simplest to understand. The main priority for day-to-day use will be to reference the information in the panel – then secondarily to customise the keys. If you hover over the right hand column icons, it would then turn into a field and you can click to edit the shortcut. This would be second in the priority order for the panel functionality.
3. To simplify your layout a bit more, you could think about showing just the grey button shortcuts exclusively and write out the word 'Shift' etc instead of using the symbol inside the button. That way you don't need the text and icon together.

Hope that helps 😁

**Reply by om-ray** [reply]
got it, ill try and get those changes into the design

**Reply by om-ray** [reply]
After reading your suggestions and looking at the other examples you provided, I came up with this new design.
![Group 31](https://user-images.githubusercontent.com/38233712/167873383-97dcc275-5e83-4a70-ab49-0355021b0d2b.png)

**Reply by samejr** [reply]
Hey @123om123 this is looking much cleaner. I've got a few more pointers for you to hopefully clean it up even more:

1. Rather than repeating the subtitles under the main titles in each section, I would just remove them completely. I think it's clear enough without this extra info. So I would remove the words "Shortcut" and "Current Hotkey" in each section.
2. When you need to press a combination of keys, I would separate the key icons into separate buttons like this:
<img width="394" alt="CleanShot 2022-05-12 at 09 21 05@2x" src="https://user-images.githubusercontent.com/7555566/168028063-73d945eb-e49a-4aca-9a78-849e3b3a0cec.png">

And when you need to show the "Cmd" and "Ctrl" key, I would write "or" between them like the example below. Then you can remove the "+" between keys completely.
<img width="394" alt="CleanShot 2022-05-12 at 09 20 52@2x" src="https://user-images.githubusercontent.com/7555566/168028197-4e8b24fa-d029-4a22-ae82-bbb5fb639cd6.png">

3. A smaller point but one of my general typography rules is that I only abbreviate words if it's absolutely necessary, like in the instance there isn't enough screen real estate. So in your example above I would spell out "Miscellaneous" in full. Or if you wanted a shorter word, use "Other".

**Reply by om-ray** [reply]
ahh that makes alot of sense, ill go change those things!

**Reply by om-ray** [reply]
![Frame 1](https://user-images.githubusercontent.com/38233712/168068541-3cb73fe7-24eb-414d-8ea6-c74fa9aa3fcd.png)
@samejr Heres the new design, i got rid of the repeating subheadings and aligned the buttons to the right like the invision freehand example, I also got rid of the "+"'s and fixed a mistake with one of the shortcuts

**Reply by samejr** [reply]
Hey @123om123 this is much cleaner now, nice work. Just a few copy amends now I think before it's worth building into the app:

1. I think remove 'History' section all together and add 'Back in history' and 'Forward in history' to the Navigation section. Like the Slack App example above.
2. Rename the 'Search Bar' heading to just 'Search' and move the 'Open Search' item from 'Other' to the top of 'Search'.
3. Move 'Reset Path' into the Navigation section
4. Change 'Copy Node' to 'Copy current selected node'
5. Change references to 'Hotkeys' into 'Shortcuts' as it seems to be more of the convention to call them shortcuts.

Other than those copy changes, would you be up for building this feature into the app?

**Reply by om-ray** [reply]
@samejr 
Ah, that makes sense ill work on those changes!

> Other than those copy changes, would you be up for building this feature into the app?

Absolutely! id love to add this to the app

**Reply by om-ray** [reply]
![Frame 1](https://user-images.githubusercontent.com/38233712/168861343-d8d8acde-cc78-4053-a241-9c2a6d2e22e7.svg)
This is the new design with the changes you recommended.

**Reply by samejr** [reply]
This looks great! 
Please do take a swing at implementing it – here are a few pointers:
- Check the project for existing components to create the key icons. They already exist in the project so it'd be good to modify (if needed) and reuse them.
- You might like to use a [Radix](https://www.radix-ui.com) component for the UI
- Include a text link in the bottom left called 'Shortcuts' that toggles the panel. Remove all the other shortcut text in that bottom panel too. You'll want to consider the styling of this button in the same look/feel as the rest of the app.
- Consider clicking the background to dismiss the panel and an Esc to close shortcut.
- Open a PR when you're ready and me or someone else will check it out and provide feedback.

**Reply by om-ray** [reply]
@samejr Thank you so much for the advice! I'm currently having some trouble with the local version of the app (I can't open JSON files) and @ericallam is looking into that now, but I'm still working on it by making the react component in a separate project just so I can see how it all looks

**Reply by samejr** [reply]
No problem at all - and great that you can make a start anyway. Does the app run locally but the JSON you paste in doesn't work? Or does the whole app just not booting up locally? If it helps I added some local JSON files in a folder called 'Examples' which I know work.

**Reply by om-ray** [reply]
@samejr The problem got fixed and I'm currently working on it!

**Reply by samejr** [reply]
Fantastic!

**Comment by om-ray** [comment]
<img width="1778" alt="Screen Shot 2022-05-19 at 12 45 16 PM" src="https://user-images.githubusercontent.com/38233712/169232831-73023e04-68f7-47d1-aafb-c6a8da271e6e.png">
@samejr This is how the app is looking currently with the panel open (ignore the blue line around the panel, i just need to add outline:none in the CSS) it closes when you press escape or when you click outside of the modal

**Reply by samejr** [reply]
Looking nice! Couple of tweaks:
1. I think you should be able to close the panel with an X in the top right.
2. I would remove the top sentence "View and change..." for now until that feature is built in.
3. You need to add the PC shortcut any time you use 'Opt'. So should say 'Opt' or 'Alt' 'S' for e.g.
4. Your bottom shortcut says 'Toggle Shortcuts Panel' which would imply that I can close this panel with Opt S. Is that true?
5. Don't forget dark mode styling if you haven't already added it.
6. Simplify the bottom left footer link by just having it say "Shortcuts" "Opt" "/" "Alt" "S"

**Reply by matt-aitken** [reply]
One small thing. Ideally we would only display the relevant shortcut keys for the device the user is on. So, if you're on Windows we'd show Ctrl/Alt and on Mac Cmd/Option.

**Reply by samejr** [reply]
Hey @123om123 just checked out your demo video. It's looking really nice. Can I suggest you create 2 separate PRs for this feature. One which is a simple panel displaying the shortcuts. Then another which adds the ability to edit those shortcuts? That way it will be easier to give feedback and get the simple version looking/functioning great first. Hope that makes sense.

**Reply by om-ray** [reply]
@samejr

> I think you should be able to close the panel with an X in the top right.

I'm currently in the process of adding that 😄 

> I would remove the top sentence "View and change..." for now until that feature is built in.

Ah that makes sense, I'll get rid of that

> You need to add the PC shortcut any time you use 'Opt'. So should say 'Opt' or 'Alt' 'S' for e.g.

Ill add that right now

> Your bottom shortcut says 'Toggle Shortcuts Panel' which would imply that I can close this panel with Opt S. Is that true?

Yep, you can close and open the panel with Opt s

>Don't forget dark mode styling if you haven't already added it.

I had completely forgotten about that! Thank you for reminding me!

**Reply by om-ray** [reply]
> Hey @123om123 just checked out your demo video. It's looking really nice. Can I suggest you create 2 separate PRs for this feature. One which is a simple panel displaying the shortcuts. Then another which adds the ability to edit those shortcuts? That way it will be easier to give feedback and get the simple version looking/functioning great first. Hope that makes sense.

Thats a great idea! ill do that!

**Reply by samejr** [reply]
> One small thing. Ideally we would only display the relevant shortcut keys for the device the user is on. So, if you're on Windows we'd show Ctrl/Alt and on Mac Cmd/Option.

I agree with this – @123om123 can you try to detect the platform and present only the relevant shortcut? The user experience will be a lot better for it.

**Reply by om-ray** [reply]
> One small thing. Ideally we would only display the relevant shortcut keys for the device the user is on. So, if you're on Windows we'd show Ctrl/Alt and on Mac Cmd/Option.

That makes a lot of sense, Ill do some research on how to do that and implement it!

**Reply by om-ray** [reply]
@samejr I'm creating a pull request for the panel itself without the inputs

**Reply by om-ray** [reply]
> One small thing. Ideally we would only display the relevant shortcut keys for the device the user is on. So, if you're on Windows we'd show Ctrl/Alt and on Mac Cmd/Option.

@matt-aitken If the users OS is linux should we display both possibilities?

**Reply by matt-aitken** [reply]
Let's show Windows shortcuts in that situation

**Reply by om-ray** [reply]
> Let's show Windows shortcuts in that situation

got it!

**Reply by om-ray** [reply]
@samejr How do the themes work? I cant seem to figure it out

**Reply by samejr** [reply]
I've not checked out your PR yet so you've probably already worked this out but we use Tailwind utility classes for all our CSS. Tailwind defines dark mode classes by prefixing 'dark:' Check out one of our other component for an example - or the Tailwind documentation for dark mode is here: https://tailwindcss.com/docs/dark-mode

**Reply by om-ray** [reply]
@samejr well i used styled components from stitches for a lot of the styling, how would that work with tailwind?

**Reply by om-ray** [reply]
@samejr I figured it out and i think it looks pretty good! although I am running into a problem where as soon as I switch to dark mode the application crashes, so I have to downgrade to node 16 to fix that

**Comment by om-ray** [comment]
@samejr This is the all new shortcut panel in dark mode!
<img width="1774" alt="Screen Shot 2022-05-19 at 10 41 58 PM" src="https://user-images.githubusercontent.com/38233712/169358645-064a5a55-6fac-4463-99ed-c419f5f97bed.png">

**Comment by om-ray** [comment]
@samejr 
https://github.com/jsonhero-io/jsonhero-web/pull/67
This is the PR for the Shortcut panel

**Comment by om-ray** [comment]
@samejr @ericallam this is the PR for the Custom keyboard shortcuts https://github.com/jsonhero-io/jsonhero-web/pull/68
